### ISSUE TITLE
💫 Idea: Learn a compressed representation of the word vectors

### ISSUE THREAD

**Issue by honnibal** [explosion/spaCy#508]
I'd like to switch the default word vector data to the Common Crawl vectors, but even with the 1m token cutoff they're a bit big. It'd be nice to learn a compressed form of them. I think this would be a nice experimental task for someone to take on. The experiment really has nothing to do with spaCy specifically — I would do it with Keras (or maybe Tensor Flow directly).

There are lots of potential approaches here. Here's one I think will work well.

Train a model to minimise  `freqs[word_id] * cosine(small_vectors[word_id], original_vectors[word_id])`. Note that it's very important that the objective should be sensitive to the word frequency, because we want to minimise the cosine loss over _tokens_, not _types_. We care much less about the fidelity of the vectors that will be assigned to rare words.

Okay, so how should the model be structured? The simplest thing to try is a bottle-neck architecture. Let's say we start with vectors of length 300. We can learn a projection layer that maps the vectors into a smaller dimensionality --- let's say 128 dimensions. Then you have a layer that maps from 128 back to 300 dimensions, so that we can do our cosine loss. We're trying to learn a mapping that preserves the information in the most common words. Once we've trained our model, we can save only the 128-dimension values for each word in our vocabulary, and if we want, the output projection matrix.

I think we can do better than the bottle-neck architecture, though. The problem with the bottle-neck architecture is that it's compressing the wrong dimension of the word vector matrix. Instead of compressing the length of every vector, I think it will be better to try to collapse similar vectors together. I'm sure there's prior art that achieves exactly this very well.

Here's a speculative suggestion, using something like the well-known feature hashing trick. What if we had the following model:

`vector[word_id] = (table1[hash1(word_id) % table1.shape[0]] * weights1[word_id]) + (hash2(word_id) % table2.shape[0]] * weights2[word_id])`.

Let's say we start with an embedding table of shape `(1000000, 300)`, i.e. we have 300 dimensional embeddings for a vocabulary of 1m words. We learn two vector tables, `table1` and `table2`, each of shape `(5000, 300)`, and two weight vectors, each of shape `(1000000,)`. Assume for now that the functions `hash1` and `hash2` are non-parametric, i.e. we're not learning these hash functions. They just distribute the word IDs into the two tables.

The idea here is that we care a lot about the fidelity of the vectors for common words, so we want to make it easy to learn a model that doesn't distort these words. The model above makes it easy to represent vectors for the top 10000 words, and ignore all others. I think that's a competitive baseline for compressing the vectors, so it's good to have a model where that solution should be easy to find.

**Comment by shirish93** [comment]
Here's a relevant paper that might be useful:

http://papers.nips.cc/paper/4831-angular-quantization-based-binary-codes-for-fast-similarity-search

I've looked around for relevant implementations and not found any reasonable ones. The first part of the paper is pretty easy to implement, but I can't quite wrap my head around on the part that involves preserving cosine similarity.

Is this the sort of idea you were looking for?

**Comment by transfluxus** [comment]
Is this the default behaviour now? I'm wondering would it be possible to compress an existing embedding?
